isisd/isis_main.c

```diff
@@ -196,6 +196,93 @@ FRR_DAEMON_INFO(isisd, ISIS, .vty_port = ISISD_VTY_PORT,
 		.n_yang_modules = array_size(isisd_yang_modules),
 );
 
+#ifdef FUZZING
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+
+static bool FuzzingInit(void)
+{
+	const char *name[] = {"isisd"};
+
+#ifdef FABRICD
+	frr_preinit(&fabricd_di, 1, (char **)name);
+#else
+	frr_preinit(&isisd_di, 1, (char **)name);
+#endif
+
+	/* INIT */
+	isis_master_init(frr_init_fast());
+
+	isis_error_init();
+	access_list_init();
+	access_list_add_hook(isis_filter_update);
+	access_list_delete_hook(isis_filter_update);
+	isis_vrf_init();
+	prefix_list_init();
+	prefix_list_add_hook(isis_prefix_list_update);
+	prefix_list_delete_hook(isis_prefix_list_update);
+	isis_init();
+	isis_circuit_init();
+	isis_spf_init();
+	isis_route_map_init();
+	isis_mpls_te_init();
+	isis_sr_init();
+	lsp_init();
+	isis_zebra_init(master, 1);
+	isis_bfd_init(master);
+
+	return true;
+}
+
+static struct isis_circuit *FuzzingCreateIsis(void)
+{
+	struct prefix p;
+	struct interface *ifp = NULL;
+	struct connected *conn = NULL;
+	struct isis_area *area = NULL;
+	struct isis_circuit *c = NULL;
+
+	ifp = if_create_ifindex(69, 0);
+	ifp->mtu = 68;
+	str2prefix("11.0.2.2/24", &p);
+
+	conn = connected_add_by_prefix(ifp, &p, NULL);
+
+	c = isis_circuit_new(ifp, "fuzz-tag");
+
+	area = isis_area_create("FUZZ_AREA", NULL);
+	isis_area_add_circuit(area, c);
+
+	return c;
+}
+
+static struct isis_circuit *FuzzingIsis;
+static bool FuzzingInitialized;
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
+{
+	struct isis_circuit *c;
+	static uint8_t ssnpa[6] = {0x2c, 0x54, 0x91, 0x88, 0xc9, 0xe3};
+
+	if (!FuzzingInitialized) {
+		FuzzingInit();
+		FuzzingInitialized = true;
+		FuzzingIsis = FuzzingCreateIsis();
+	}
+
+	c = FuzzingIsis;
+
+	c->rcv_stream = stream_new(MAX(1, size));
+
+	stream_put(c->rcv_stream, data, size);
+
+	isis_handle_pdu(c, ssnpa);
+
+	return 0;
```

Comment:
Need to free the stream right? Or is it consumed by `isis_handle_pdu`?

Reply:
`isis_handle_pdu` doesn't call `stream_free` so I added a call to it. That results in valgrind reporting less bytes as "still reachable". Interestingly, the ospfd fuzzer doesn't free the input stream it uses either.

I'm not sure if this could affect fuzzing performance as even after this change valgrind still reports a lot of bytes as "still reachable". I did some simple testing, and so far there doesn't seem to be a difference on my setup. I guess that isn't surprising as the stream is around a few hundred bytes. Maybe it would be more significant if AFL persistent mode was used.

Reply:
Not perf, I just wanted to make sure we didn't have a memory leak here, cause that will show up in our automated fuzz results in oss-fuzz.